File: ru.kmorozov.gbd.desktop/src/test/java/ru/kmorozov/gbd/core/test/ExtractorTest.java
package ru.kmorozov.gbd.core.test;

import org.apache.commons.io.IOUtils;
import org.eclipse.jetty.http.MimeTypes.Type;
import org.eclipse.jetty.server.Request;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.DefaultHandler;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;
import ru.kmorozov.gbd.core.config.GBDOptions;
import ru.kmorozov.gbd.core.config.IGBDOptions;
import ru.kmorozov.gbd.core.config.storage.AbstractContextProvider;
import ru.kmorozov.gbd.core.logic.context.ExecutionContext;
import ru.kmorozov.gbd.core.logic.extractors.google.GoogleBookInfoExtractor;
import ru.kmorozov.gbd.core.logic.extractors.shpl.ShplBookExtractor;
import ru.kmorozov.gbd.core.logic.library.metadata.ShplMetadata;
import ru.kmorozov.gbd.core.logic.output.consumers.DummyBookInfoOutput;
import ru.kmorozov.gbd.desktop.library.OptionsBasedProducer;
import ru.kmorozov.gbd.desktop.library.SingleBookProducer;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.Charset;

import static org.hamcrest.CoreMatchers.*;

/**
 * Created by sbt-morozov-kv on 16.11.2016.
 */
public class ExtractorTest {

    private static final String SHPL_HTML_RESOURCE = "shpl/book.html";
    private static final String GOOGLE_HTML_RESOURCE = "google/book.html";

    @Before
    public void initServer() {
        ExecutionContext.initContext(new DummyBookInfoOutput(), true);
        final IGBDOptions mockOptions = Mockito.mock(IGBDOptions.class);
        Mockito.when(mockOptions.getOutputDir()).thenReturn("E:\\Work\\gbd\\");
        Mockito.when(mockOptions.getProxyListFile()).thenReturn("E:\\Work\\gbd\\proxy.txt");
        GBDOptions.init(mockOptions);

        final Server server = new Server(80);
        server.setHandler(new DefaultHandler() {

            @Override
            public void handle(final String target, final Request baseRequest, final HttpServletRequest request, final HttpServletResponse response) throws IOException {
                response.setStatus(200);
                response.setContentType(Type.TEXT_HTML.toString());

                String resFileName = null;
                if (request.getRequestURI().contains(ShplMetadata.SHPL_BASE_URL)) resFileName = SHPL_HTML_RESOURCE;
                else if (request.getRequestURI().contains("google")) resFileName = GOOGLE_HTML_RESOURCE;

                if (null == resFileName) return;

                final File res = new File(getClass().getClassLoader().getResource(resFileName).getFile());
                final String data = IOUtils.toString(new FileInputStream(res), Charset.forName("UTF-8"));
                try (OutputStream os = response.getOutputStream()) {
                    os.write(data.getBytes());
                }
            }
        });

        try {
            server.start();
        } catch (final Exception e) {
            e.printStackTrace();
        }
    }

    @Test
    public void shplBookInfoTest() {
        final ShplBookExtractor extractor = new ShplBookExtractor("http://localhost");

        Assert.assertThat(extractor.getBookInfo(), notNullValue());
    }

    @Test
    public void bookContextLoadTest() {
        final AbstractContextProvider contextProvider = AbstractContextProvider.getContextProvider();

        final int ctxSizeBefore = contextProvider.getContextSize();
        Assert.assertTrue(0 < ctxSizeBefore);

        ExecutionContext.INSTANCE.addBookContext(new OptionsBasedProducer(), null, null);
        ExecutionContext.INSTANCE.addBookContext(new SingleBookProducer("http://localhost/elib.shpl.ru"), null, null);

        contextProvider.updateContext();
        contextProvider.refreshContext();

        final int ctxSizeAfter = contextProvider.getContextSize();
        Assert.assertThat(ctxSizeAfter, is(ctxSizeBefore + 1));
    }

    @Test
    public void googleExtractorTest() {
        final GoogleBookInfoExtractor extractor = new GoogleBookInfoExtractor("test") {
            @Override
            protected String getBookUrl() {
                return "http://localhost/google";
            }
        };

        Assert.assertThat(extractor.getBookInfo(), notNullValue());
    }
}

File: ru.kmorozov.gbd.desktop/src/test/java/ru/kmorozov/gbd/core/test/ExtractorTest.kt
package ru.kmorozov.gbd.core.test

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.eclipse.jetty.http.MimeTypes.Type
import org.eclipse.jetty.server.Request
import org.eclipse.jetty.server.Server
import org.eclipse.jetty.server.handler.DefaultHandler
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.notNullValue
import org.hamcrest.MatcherAssert
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import ru.kmorozov.db.core.logic.model.book.BookInfo
import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.core.config.IGBDOptions
import ru.kmorozov.gbd.core.config.IStorage
import ru.kmorozov.gbd.core.loader.LocalFSStorage
import ru.kmorozov.gbd.core.logic.context.ContextProvider
import ru.kmorozov.gbd.core.logic.context.ExecutionContext
import ru.kmorozov.gbd.core.logic.extractors.google.GoogleBookInfoExtractor
import ru.kmorozov.gbd.core.logic.extractors.shpl.ShplBookInfoExtractor
import ru.kmorozov.gbd.core.logic.library.metadata.ShplMetadata
import ru.kmorozov.gbd.core.producers.OptionsBasedProducer
import ru.kmorozov.gbd.core.producers.SingleBookProducer
import ru.kmorozov.gbd.logger.output.ReceiverProvider
import java.io.File
import java.io.IOException
import java.nio.file.Files

/**
 * Created by sbt-morozov-kv on 16.11.2016.
 */
class ExtractorTest {

    @Before
    fun initServer() {
        ExecutionContext.initContext(ReceiverProvider.getReceiver(), true)
        val mockOptions = Mockito.mock(IGBDOptions::class.java)
        Mockito.`when`<IStorage>(mockOptions.storage).thenReturn(LocalFSStorage.getStorage("E:\\Work\\gbd\\"))
        Mockito.`when`(mockOptions.proxyListFile).thenReturn("E:\\Work\\gbd\\proxy.txt")
        GBDOptions.init(mockOptions)

        val server = Server(80)
        server.handler = object : DefaultHandler() {

            @Throws(IOException::class)
            override fun handle(target: String, baseRequest: Request, request: HttpServletRequest, response: HttpServletResponse) {
                response.status = 200
                response.contentType = Type.TEXT_HTML.toString()

                var resFileName: String? = null
                if (request.requestURI.contains(ShplMetadata.SHPL_BASE_URL))
                    resFileName = SHPL_HTML_RESOURCE
                else if (request.requestURI.contains("google")) resFileName = GOOGLE_HTML_RESOURCE

                if (null == resFileName) return

                val res = File(javaClass.classLoader.getResource(resFileName)!!.file)
                val data = Files.readString(res.toPath())
                response.outputStream.use { os -> os.write(data.toByteArray()) }
            }
        }

        try {
            server.start()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    @Test
    fun shplBookInfoTest() {
        val extractor = ShplBookInfoExtractor("http://localhost")

        MatcherAssert.assertThat<BookInfo>(extractor.bookInfo, notNullValue())
    }

    @Test
    fun bookContextLoadTest() {
        val contextProvider = ContextProvider.contextProvider

        val ctxSizeBefore = contextProvider.contextSize
        Assert.assertTrue(0 < ctxSizeBefore)

        ExecutionContext.INSTANCE.addBookContext(OptionsBasedProducer(), null!!)
        ExecutionContext.INSTANCE.addBookContext(SingleBookProducer("http://localhost/elib.shpl.ru"), null!!)

        contextProvider.updateContext()

        val ctxSizeAfter = contextProvider.contextSize
        MatcherAssert.assertThat(ctxSizeAfter, `is`(ctxSizeBefore + 1))
    }

    @Test
    fun googleExtractorTest() {
        val extractor = object : GoogleBookInfoExtractor("test") {
            override val bookUrl: String
                get() = "http://localhost/google"
        }

        MatcherAssert.assertThat<BookInfo>(extractor.bookInfo, notNullValue())
    }

    companion object {

        private const val SHPL_HTML_RESOURCE = "shpl/book.html"
        private const val GOOGLE_HTML_RESOURCE = "google/book.html"
    }
}
